### PR TITLE
compaction: Perform integrity checks on compacting SSTables

### PR DESCRIPTION
**Pull request by ndragazis**
This PR enables compaction tasks to verify the integrity of the input data through checksum and digest checks. The mechanism for integrity checking was introduced in previous PRs (#20207, #20720) as a built-in functionality of the input streams. This PR integrates this mechanism with compaction. The change applies to all compaction types and covers both compressed and uncompressed SSTables adhering to the 3.x format. If a compaction task reads only part of an SSTable, then only the per-chunk checksums are verified, not the digest.

The PR consists of:
* Changes to mx readers to support integrity checking. The kl readers, considered as compatibility-only, were left unchanged. Also, integrity checking on single-partition reversed reads (`data_consume_reversed_partition()`) remains unsupported by mx readers as this is not used in compaction.
* Changes to `sstable` and `sstable_set` APIs to allow toggling integrity checks for mx readers.
* Activation of integrity checking for all compaction types.
* Tests for all compaction types with corrupted SSTables.

Integrity checks come at a cost. For uncompressed SSTables, the cost is the loading of the CRC and Digest components from disk, and the calculation of checksums and digest from the actual data. For compressed SSTables, checksums are stored in-place and they are being checked already on all reads, so the only extra cost is the loading and calculation of the digest. The measurements show a ~5% regression in compaction performance for uncompressed SSTables, and a negligible regression for compressed SSTables.

Command: `perf-sstable --smp=1 --cpuset=1 --poll-mode --mode=compaction --iterations=1000 --partitions 10000 --sstables=1 --key_size=4096 --num_columns=15 --column_size={32, 1024, 3500, 7000, 14500}`

Uncompressed SSTables:
```
+--------------+-----------------------+----------------------+------------+
| SSTable Size | No Integrity (p/sec)  | Integrity (p/sec)    | Regression |
+--------------+-----------------------+----------------------+------------+
| 50  MiB      | 65175.59 +- 80.82     | 61814.63 +- 72.88    | 5.16%      |
| 200 MiB      | 41795.10 +- 60.39     | 39686.28 +- 45.05    | 5.05%      |
| 500 MiB      | 21087.41 +- 30.72     | 20092.93 +- 25.05    | 4.72%      |
| 1   GiB      | 12781.64 +- 21.77     | 12233.94 +- 21.71    | 4.29%      |
| 2   GiB      |  6629.99 +-  9.40     |  6377.13 +-  8.28    | 3.81%      |
+--------------+-----------------------+----------------------+------------+
```
Compressed SSTables:
```
+--------------+-----------------------+----------------------+------------+
| SSTable Size | No Integrity (p/sec)  | Integrity (p/sec)    | Regression |
+--------------+-----------------------+----------------------+------------+
| 50  MiB      | 53975.05 +- 63.18     | 53825.93 +- 62.28    |  0.28%     |
| 200 MiB      | 28687.94 +- 26.58     | 28689.41 +- 26.91    |  0%        |
| 500 MiB      | 13865.35 +- 15.50     | 13790.41 +- 14.88    |  0.54%     |
| 1   GiB      |  7858.10 +-  7.71     |  7829.75 +-  9.66    |  0.36%     |
| 2   GiB      |  4023.11 +-  2.43     |  4010.54 +-  2.55    |  0.31%     |
+--------------+-----------------------+----------------------+------------+
(p/sec = partitions/sec)
```

Refs #19071.

New feature, no backport is needed.

### CODE REVIEW

**Comment by avikivity** [comment]
> Activation of integrity checking for all compaction types.

What's the performance impact?

**Comment by denesb** [comment]
> Refs #19071.

Why only refs? Should be "Fixes", no?

File: sstables/mx/reader.cc [hunk 1]
```diff
@@ -1781,13 +1797,16 @@ class mx_sstable_full_scan_reader : public mp_row_consumer_reader_mx {
...
+        if (!is_initialized()) {
+            co_await maybe_initialize();
         }
         if (_context->eof()) {
             _end_of_stream = true;
-            return make_ready_future<>();
+            co_return;
         }
-        return _context->consume_input();
+        co_await _context->consume_input();
```
Comment (denesb) [inline sstables/mx/reader.cc:1828]: In general `fill_buffer()` is hot path for readers, so we don't use co-routines here, because they add an unconditional allocation.
One could argue that this doesn't apply to `full_scan_reader`, because it is not used in any performance-sensitive code-paths, but I think it is worth keeping all reader's `fill_buffer()` coroutine-free.
There is a simple workaround to keep the code relatively straightforward:
```c++
future<> fill_buffer() {
   if (!is_initialized()) {
       return initialize().then([this] { return fill_buffer(); };
    }
    // rest of the code for `fill_buffer()`.
```

This trick allows for saving an indentation level, at the expense of a recursion and this is considered quite a standard practice for such cases.

Comment (denesb) [reply sstables/mx/reader.cc:1828]: Note: it is fine to use coroutines in `initialize()` because it is called once in the lifetime of the reader.

File: sstables/mx/reader.cc [hunk 2]
```diff
@@ -1764,8 +1766,22 @@ class mx_sstable_full_scan_reader : public mp_row_consumer_reader_mx {
...
+        return bool(_context);
+    }
+    future<> maybe_initialize() {
+        if (is_initialized()) {
+            co_return;
+        }
+        if (_integrity) {
+            // Caller must retain a reference to checksum component while in use by the stream.
+            _checksum = co_await _sst->read_checksum();
+            co_await _sst->read_digest();
```
Comment (denesb) [inline sstables/mx/reader.cc:1801]: Why not make this conditional? Maybe caller (or a parallel scrub) already loaded the component. Or `read_checksum()` and `read_digest()` already has this optimization (no-op if already loaded)?

**Review by denesb — APPROVED** [review]
Very well structured series with detailed commitlogs, very easy to review. Well done.

I have one comment and we need a performance test as @avikivity requested, otherwise LGTM.

**Comment by ndragazis** [comment]
>What's the performance impact?

Don't know yet. I will come back with some numbers.

**Comment by ndragazis** [comment]
> > Refs https://github.com/scylladb/scylladb/issues/19071.
>
> Why only refs? Should be "Fixes", no?

The issue proposes adding checks to repair and streaming as well. This is not covered in this PR.

**Comment by denesb** [comment]
> > > Refs #19071.
> > 
> > 
> > Why only refs? Should be "Fixes", no?
> 
> The issue proposes adding checks to repair and streaming as well. This is not covered in this PR.

Right, missed that.

**Comment by raphaelsc** [comment]
@ndragazis what happens when compaction (e.g. cleanup compaction) skips some chunks or reads only a portion of a given chunk? do we skip the full digest check on completion? we don't have digest check for other sstable components yet, right?

File: sstables/mx/reader.cc (hunk 1 above)
Comment (ndragazis) [reply sstables/mx/reader.cc:1828]: Ack, I wasn't aware of the allocation overhead. I just wanted to make the code more readable.

File: sstables/mx/reader.cc (hunk 2 above)
Comment (ndragazis) [reply sstables/mx/reader.cc:1801]: Both functions are idempotent. They act as no-ops if the components have already been loaded.

Comment (denesb) [reply sstables/mx/reader.cc:1801]: Great, that is actually better than having all their callers have to do this.

**Comment by ndragazis** [comment]
> @ndragazis what happens when compaction (e.g. cleanup compaction) skips some chunks or reads only a portion of a given chunk? do we skip the full digest check on completion?

Yes, that's the first patch in the series. We skip the digest check, but we continue checking the checksums of the chunks that are being read.

> we don't have digest check for other sstable components yet, right?

Right, but we have an issue to maintain digests for all components, see: https://github.com/scylladb/scylladb/issues/20100.

**Comment by ndragazis** [comment]
> >What's the performance impact?
>
>Don't know yet. I will come back with some numbers.

TL;DR: There is a 5-8% drop for small uncompressed SSTables. For large uncompressed SSTables (2GiB), there is a weird increase in performance that seems to appear consistently on my measurements. For compressed SSTables, the differences are generally smaller.

Note: I was noticing significant flunctuations in the mean values of my measurements, so I ran many of them multiple times and took the average. Not sure where this comes from.

## Compaction Performance by SSTable size

Configuration:
* `build/release/scylla perf-sstable --testdir=./perf-tests --mode=compaction --smp=8 --iterations=10 --partitions 40000 --sstables=4 --key_size=4096 --num_columns=15 --column_size=<SIZE>`
* Column sizes: 32, 1024, 3500, 7000, 14500

Compaction w/ Uncompressed SSTables:
```
+--------------+-----------------------+----------------------+------------+
| SSTable Size | No Integrity (p/sec)  | Integrity (p/sec)    | Regression |
+--------------+-----------------------+----------------------+------------+
| 50  MiB      | 156730.23 +- 1100.33  | 150221.41 +- 2144.56 |   4.15%    |
| 200 MiB      |  44669.62 +- 2554.83  |  41478.58 +- 3464.16 |   7.14%    |
| 500 MiB      |  10807.82 +- 1365.89  |  10706.87 +- 1119.50 |   0.93%    |
| 1   GiB      |   4634.28 +-  466.45  |   4325.60 +-  334.78 |   6.66%    |
| 2   GiB      |   2045.59 +-  157.62  |   2132.86 +-  228.72 |  -4.27%    |
+--------------+-----------------------+----------------------+------------+
p/sec = partitions/sec
```
Compaction w/ Compressed SSTables:
```
+--------------+-----------------------+----------------------+------------+
| SSTable Size | No Integrity (p/sec)  | Integrity (p/sec)    | Regression |
+--------------+-----------------------+----------------------+------------+
| 50  MiB      | 143869.43 +- 3056.16  | 142328.04 +- 2765.10 |   1.07%    |
| 200 MiB      |  47859.97 +- 3737.93  |  43936.32 +- 2900.32 |   8.20%    |
| 500 MiB      |  11252.03 +- 1615.93  |  11546.52 +- 1701.46 |  -2.62%    |
| 1   GiB      |   4504.07 +-  441.25  |   4413.01 +-  450.14 |   2.02%    |
| 2   GiB      |   1984.26 +-  118.37  |   1920.44 +-   52.99 |   3.22%    |
+--------------+-----------------------+----------------------+------------+
```

## Compaction Performance by number of SSTables

Configuration:
* `./build/release/scylla perf-sstable --testdir=./perf-tests --mode=compaction --smp=8 --iterations=10 --partitions=<PARTITIONS> --sstables=<SSTABLES> --key_size=4096`
* 10k partitions per SSTable, i.e. `SSTABLES * 10000` 

Compaction w/ Uncompressed SSTables:
```
+----------+-----------------------+----------------------+------------+
| SSTables | No Integrity (p/sec)  | Integrity (p/sec)    | Regression |
+----------+-----------------------+----------------------+------------+
|  1       | 445608.61 +- 3241.45  | 419109.59 +- 5323.71 |  5.95%     |
|  2       | 232529.97 +- 2357.60  | 215685.48 +- 4370.87 |  7.24%     |
|  4       | 145806.09 +- 3809.44  | 137756.42 +- 3800.69 |  5.52%     |
|  8       |  81957.55 +- 1792.55  |  78038.62 +-  835.34 |  4.78%     |
|  16      |  43961.80 +-  882.26  |  41258.13 +-  684.44 |  6.15%     |
|  32      |  22163.39 +-  500.06  |  20371.58 +-  365.66 |  8.08%     |
+----------+-----------------------+----------------------+------------+
p/sec = partitions/sec
```
Compaction w/ Compressed SSTables:
```
+----------+-----------------------+----------------------+------------+
| SSTables | No Integrity (p/sec)  | Integrity (p/sec)    | Regression |
+----------+-----------------------+----------------------+------------+
|  1       | 403271.61 +- 3771.31  | 406423.92 +- 5075.59 |  -0.78%    |
|  2       | 218162.17 +-  944.82  | 218537.68 +- 1527.16 |  -0.17%    |
|  4       | 137886.77 +- 3686.16  | 136444.90 +- 4822.54 |   1.05%    |
|  8       |  77619.91 +- 1351.74  |  77617.44 +- 1786.50 |   0%       |
|  16      |  41429.95 +-  792.98  |  42814.06 +-  997.22 |  -3.34%    |
|  32      |  20607.62 +-  653.16  |  21033.39 +-  443.94 |  -2.07%    |
+----------+-----------------------+----------------------+------------+
```

**Comment by avikivity** [comment]
The regression appears quite large, but are those numbers trustworthy? Why should performance improve for 32 sstables? Why are the numbers for 500MB and 1GB different? Seems very random.

**Comment by raphaelsc** [comment]
> The regression appears quite large, but are those numbers trustworthy? Why should performance improve for 32 sstables? Why are the numbers for 500MB and 1GB different? Seems very random.

It's possible some unrelated process causes the noise if running on same vcpus. I suggest re-running with --smp 1 and pinning the process to some unbusy vcpu (using scylla option --cpuset). Close browser, etc.

**Comment by ndragazis** [comment]
In an effort to reduce the noise in my results, I made the following changes:
* Compact 1 SSTable instead of many.
* Use 1 shard.
* Pin the shard to the same core for all measurements.
* Set the cpu governor to "powersave". This stabilizes frequency at 2.5GHz and there is no throttling.
* Use `--poll-mode`.
* Run over tmpfs.
* Increase number of iterations.

command: `build/release/scylla perf-sstable --testdir=/var/lib/scylla --smp=1 --cpuset=1 --poll-mode --mode=compaction --iterations=1000 --partitions 10000 --sstables=1 --key_size=4096 --num_columns=15 --column_size=...`

The numbers got more stable this way, but I would still experience flunctuations in the mean values; same configuration would produce different values and the diff would be higher than the error rates.

I thought that maybe the culprit is the SSTables, which are randomly generated.

As a next step, I ran the before-and-after measurements with the exact same SSTables. This is not supported by the tool, so here is how I did it:
1. Run the tool with gdb and add a breakpoint at `sstables::compact_sstables()`.
2. When the tool hits the breakpoint, make a copy of the SSTable that has been created.
3. Repeat 1-2 for all five test configurations.
4. Hack the tool so that it loads the SSTable from the filesystem.
5. Re-compile my two branches. There is no re-compilation between measurements. I have two separate clones.

The results for **uncompressed** SSTables are the following:
```
+--------------+-----------------------+----------------------+------------+
| SSTable Size | No Integrity (p/sec)  | Integrity (p/sec)    | Regression |
+--------------+-----------------------+----------------------+------------+
| 50  MiB      | 65175.59 +- 80.82     | 61814.63 +- 72.88    | 5.16%      | (1000 iterations)
| 200 MiB      | 41795.10 +- 60.39     | 39686.28 +- 45.05    | 5.05%      | (1000 iterations)
| 500 MiB      | 21087.41 +- 30.72     | 20092.93 +- 25.05    | 4.72%      | (500 iterations)
| 1   GiB      | 12781.64 +- 21.77     | 12233.94 +- 21.71    | 4.29%      | (300 iterations)
| 2   GiB      |  6629.99 +-  9.40     |  6377.13 +-  8.28    | 3.81%      | (300 iterations)
+--------------+-----------------------+----------------------+------------+
```
The regression is more stable now, as expected. The error rates were improved by the increased number of iterations. I suppose that the distribution has a few big outliers that get normalized better. Finally, note that I can still see some flunctuation in the mean values (e.g., a second measurement for 2GiB showed a regression of 4.24%).

**Comment by raphaelsc** [comment]
> In an effort to reduce the noise in my results, I made the following changes:
> * Compact 1 SSTable instead of many.
> * Use 1 shard.
> * Pin the shard to the same core for all measurements.
> * Set the cpu governor to "powersave". This stabilizes frequency at 2.5GHz and there is no throttling.
> * Use `--poll-mode`.
> * Run over tmpfs.
> * Increase number of iterations.
> 
> command: `build/release/scylla perf-sstable --testdir=/var/lib/scylla --smp=1 --cpuset=1 --poll-mode --mode=compaction --iterations=1000 --partitions 10000 --sstables=1 --key_size=4096 --num_columns=15 --column_size=...`
> 
> The numbers got more stable this way, but I would still experience flunctuations in the mean values; same configuration would produce different values and the diff would be higher than the error rates.
> 
> I thought that maybe the culprit is the SSTables, which are randomly generated.
> 
> As a next step, I ran the before-and-after measurements with the exact same SSTables. This is not supported by the tool, so here is how I did it:
> 1. Run the tool with gdb and add a breakpoint at `sstables::compact_sstables()`.
> 2. When the tool hits the breakpoint, make a copy of the SSTable that has been created.
> 3. Repeat 1-2 for all five test configurations.
> 4. Hack the tool so that it loads the SSTable from the filesystem.
> 5. Re-compile my two branches. There is no re-compilation between measurements. I have two separate clones.
> 
> The results for **uncompressed** SSTables are the following:
> ```
> +--------------+-----------------------+----------------------+------------+
> | SSTable Size | No Integrity (p/sec)  | Integrity (p/sec)    | Regression |
> +--------------+-----------------------+----------------------+------------+
> | 50  MiB      | 65175.59 +- 80.82     | 61814.63 +- 72.88    | 5.16%      | (1000 iterations)
> | 200 MiB      | 41795.10 +- 60.39     | 39686.28 +- 45.05    | 5.05%      | (1000 iterations)
> | 500 MiB      | 21087.41 +- 30.72     | 20092.93 +- 25.05    | 4.72%      | (500 iterations)
> | 1   GiB      | 12781.64 +- 21.77     | 12233.94 +- 21.71    | 4.29%      | (300 iterations)
> | 2   GiB      |  6629.99 +-  9.40     |  6377.13 +-  8.28    | 3.81%      | (300 iterations)
> +--------------+-----------------------+----------------------+------------+
> ```
> The regression is more stable now, as expected. The error rates were improved by the increased number of iterations. I suppose that the distribution has a few big outliers that get normalized better. Finally, note that I can still see some flunctuation in the mean values (e.g., a second measurement for 2GiB showed a regression of 4.24%).

Do I understand it right that for the compressed case there shouldn't be any noticeable regression since it's only adding a digest check on top of what we already have?

**Comment by ndragazis** [comment]
Yes, for compressed SSTables, we already had checksum checking enabled for all reads. Digest checking is the only new thing.

Results for compressed SSTables:
```
+--------------+-----------------------+----------------------+------------+
| SSTable Size | No Integrity (p/sec)  | Integrity (p/sec)    | Regression |
+--------------+-----------------------+----------------------+------------+
| 50  MiB      | 53975.05 +- 63.18     | 53825.93 +- 62.28    |  0.28%     | (1000 iterations)
| 200 MiB      | 28687.94 +- 26.58     | 28689.41 +- 26.91    |  0%        | (1000 iterations)
| 500 MiB      | 13865.35 +- 15.50     | 13790.41 +- 14.88    |  0.54%     | (500 iterations)
| 1   GiB      |  7858.10 +-  7.71     |  7829.75 +-  9.66    |  0.36%     | (300 iterations)
| 2   GiB      |  4023.11 +-  2.43     |  4010.54 +-  2.55    |  0.31%     | (300 iterations)
+--------------+-----------------------+----------------------+------------+
```
The differences are very small. It could be just noise.

I think it makes sense that digest checking has lower overhead than checksum checking. For checksums we apply the CRC32 algorithm on the whole chunk. For digests, we combine the already computed checksum with the partial digest. This happens in `checksum_combine_or_feed()`.

**Comment by avikivity** [comment]
Where does the regression come from? Can it be improved?

Perhaps `perf diff` will show.

**Comment by ndragazis** [comment]
In theory, the regression for uncompressed SSTables comes from:
* Checksum calculations.
* Digest calculation.
* Reading CRC and Digest from disk.

## Checksum calculations
This is the biggest factor. I used `perf_checksum` to measure libdeflate's CRC32 algorithm on my machine (single-core, 2.5GHz). Results:
```
+--------------+-----------------------+-------------------------+      
| Buffer Size  | CRC32 Throughput (GiB/s) | CRC32 Latency (msec) |      
+--------------+--------------------------+----------------------+      
| 50  MiB      | 13.51                    |   3.693              |      
| 200 MiB      | 12.72                    |  15.694              |      
| 500 MiB      | 12.70                    |  39.418              |      
| 1   GiB      | 12.60                    |  80.996              |      
| 2   GiB      | 12.70                    | 161.465              |      
+--------------+--------------------------+----------------------+
```
To compare, the compaction latencies as obtained from the previous results in https://github.com/scylladb/scylladb/pull/21153#issuecomment-2434934633 by dividing total partitions with throughput are the following:
```
+--------------+--------------------+-----------------+
| SSTable Size | No Integrity (ms)  | Integrity (ms)  |
+--------------+--------------------+-----------------+
| 50  MiB      |  153               |  162            |
| 200 MiB      |  239               |  252            |
| 500 MiB      |  474               |  498            |
| 1   GiB      |  782               |  817            |
| 2   GiB      | 1508               | 1568            |
+--------------+--------------------+-----------------+
```
One thing to notice is that the CRC32 latency doesn't simply add up to the compaction latency. Is this due to read-aheads?

## Digest calculation

Negligible. For each chunk, the data source combines the already-computed checksum of this chunk with the partial digest using the `fast_crc32_combine()`. `perf_checksum` shows that this function takes 11ns on my machine.

## Reading CRC and Digest from disk

Negligible. The Digest is 10 bytes. The CRC contains one 4-byte checksum for each chunk. This means that one 512-byte disk block can hold 128 checksums, which cover 128 * 64KiB = 8MiB of data.

My measurements show that, for a 50MiB SSTable, loading CRC and Digest from an NVMe SSD takes ~200us. This is three orders of magnitude smaller than the compaction latency.

**Comment by avikivity** [comment]
> One thing to notice is that the CRC32 latency doesn't simply add up to the compaction latency. Is this due to read-aheads?


Could be that CRC32 calculation thrashes the data cache and causes following code to run slower for a while.

Given the 5% regression is only for uncompressed sstables, and that it can prevent errors from spreading through the data, I think the regression is tolerable.

**Comment by denesb** [comment]
> > One thing to notice is that the CRC32 latency doesn't simply add up to the compaction latency. Is this due to read-aheads?
> 
> Could be that CRC32 calculation thrashes the data cache and causes following code to run slower for a while.
> 
> Given the 5% regression is only for uncompressed sstables, and that it can prevent errors from spreading through the data, I think the regression is tolerable.

Does this mean you are happy with this PR? If so, please queue (I already approved it).
